STM32 G4 ITM support and sample

andysan opened the pull request:
Add support for the Arm Instrumentation and Trace Macrocel (ITM) in the G4 SoC definitions. This has been tested on a ST Nucleo G431RB. This change also includes a configuration fragment that can be used to test SWO using the `samples/subsys/logging/logger` example.

This change can be tested with:
```
west build -b nucleo_g431rb zephyr/samples/subsys/logging/logger/ -DEXTRA_CONF_FILE=arm_itm_swo.conf
```
After flashing the board, connect a suitable debug probe and use the SWO viewer set to 2MHz (note that the core frequency will need to be specified, 170MHz on this particular board).

andysan committed:
dts: stm32g4: Add ITM support

The STM32 G4 series has a built-in Arm Instrumentation Trace
Macrocell. Set CONFIG_HAS_SWO to enable this.

Signed-off-by: Andreas Sandberg <andreas@sandberg.uk>

erwango commented on soc/arm/st_stm32/stm32g4/Kconfig.series:
```diff
@@ -14,5 +14,6 @@ config SOC_SERIES_STM32G4X
 	select CPU_HAS_ARM_MPU
 	select CLOCK_CONTROL_STM32_CUBE if CLOCK_CONTROL
 	select HAS_PM
+	select HAS_SWO
```
Out of curiosity, what is the use of `HAS_SWO`. I can't find any related code

andysan replied:
It's only used in `subsys/logging/backends/Kconfig.swo` as a dependency for `LOG_BACKEND_SWO`.

erwango replied:
Ok, my current working branch is not recent enough. I'll have a look once I'm done.

erwango reviewed (commented):
Thanks for this work, but I can't find related code in tree. Do you have any pointers or code sample that I can look at ?

andysan commented:
> Thanks for this work, but I can't find related code in tree. Do you have any pointers or code sample that I can look at ?

There isn't exactly an example in the tree. I have been using `samples/subsys/logging/logger/` and just setting `CONFIG_LOG_BACKEND_SWO=y` and then `CONFIG_LOG_BACKEND_SWO_FREQ_HZ=2000000` to match Cube Programmer's default frequency. You'll also need to set LOG_BACKEND_SWO_REF_FREQ_HZ if it hasn't already picked up its default from the CPU node or the `swo-ref-frequency` property in the ITM DT.

Looking at my own comment, I think there are a few more changes that I need to add to  this PR:

1. A commit that sets `swo-ref-frequency` for the board(s) I have tested if needed.
2. A commit that sets the default SWO output freq to 2MHz for the boards I have tested to ensure they match the defaults in the Cube programmer.
3. A commit that adds an `arm_itm_swo.conf` config overlay to the logger example. 

As far as I can tell, there is no separate pinmux configuration needed to use the SWO port.

Does this sounds like a good plan?

erwango commented:
> Does this sounds like a good plan?

That would be great.

andysan commented:
It's worth noting that a device tree node isn't strictly necessary to use SWO. The log backend currently uses the `HAS_SWO` config option to detect if ITM is available. The configuration system then tries to determine the frequency of the clock driving the ITM block by first looking for the ITM node, then the clock node of the CPU. It's possible, even likely, that the device tree updates in this change aren't actually needed.

andysan committed:
samples: Add an SWO logger configuration fragment

The subsys/logging/logger example can be used to demonstrate how to
use SWO for logging. Add a configuration fragment that enables the
logger's SWO backend. The default SWO frequency in this example is set
to 2 MHz to ensure a well-defined output bit rate instead of the
highest possible.

Signed-off-by: Andreas Sandberg <andreas@sandberg.uk>

andysan commented:
> > Does this sounds like a good plan?
> 
> That would be great.

OK, I have dropped the DT changes since they weren't needed. It feels a bit weird since that's how we usually detect devices, but SWO is apparently an exception. I have also added a configuration fragment that can be used for testing. The example defaults to 2MHz to ensure a well-defined frequency.

erwango reviewed (approved):
Ok for the STM32 part.
I let logging maintainer comment on the sample change

erwango commented:
@andysan Can you check at errors in CI? They might not be directly related to you change, but they will block it and should be fixed somehow. You can raise an issue if you don't have any clues on how to fix.

andysan commented:
> @andysan Can you check at errors in CI? They might not be directly related to you change, but they will block it and should be fixed somehow. You can raise an issue if you don't have any clues on how to fix.

It looks like the failing tests have been fixed on the main branch now. I have rebased and updated the PR. No changes to the PR itself though.

erwango reviewed (approved):
Approved on STM32 part.
